### DISCUSSION TITLE
Should allow Virtual entity has PK

### DISCUSSION THREAD

**Discussion by wenerme** [mikro-orm/mikro-orm#5444] [Ideas]
### Describe the bug

I want to reuse the base entity like

```ts
// @Entity({
//   tableName: 'last_rent_contract_view',
// })
@Entity({
 expression:`
   select distinct on (house_id) *
   from rent_contract
   order by house_id, end_date desc
 `
})
export class LastRentContractEntity extends RentContractEntity {
  @OneToOne(() => HouseEntity, { inversedBy: (v) => v.lastRentContract })
  house!: Rel<HouseEntity>;
}
```

> MetadataError: Virtual entity LastRentContractEntity cannot have primary key LastRentContractEntity.id

But this is the same as

```ts
@Entity({
  tableName: 'last_rent_contract_view',
})
```

MikroORM should allowed the PK on Virtual entity

### Reproduction

Add PK to Virtual Entity

### What driver are you using?

None

### MikroORM version

latest

### Node.js version

20+

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
It does not represent a real entity, there are no keys/indexes/uniques. If you want the value, just use `@Property` decorator, you can have an `id` there. It just cannot be marked as PK, as well, it is not a PK. Also I don't think that removing this validation would even help you do what you are trying to do.

If you want to create a virtual relation, you already can do that, just use `formula` (to return the FK value).

**Comment by wenerme** [comment]
@B4nan  Thanks, I will try the `formula` way. But I think the formula way can not support the dataloard pattern ? https://github.com/mikro-orm/mikro-orm/discussions/5443

**Reply by B4nan** [reply]
The dataloader integration does not work with `em.populate` at all, it's only for `Collection.load()` and `Reference.load()`. 

cc @darkbasic, to support this for `em.populate` we need the stuff you have in https://github.com/darkbasic/mikro-orm-dataloaders as a prerequisite I guess?

**Reply by darkbasic** [reply]
@B4nan only to a certain extent. If I recall correctly `em.populate` accepts just an array of entities with no further options and/or filters (except for the populate option itself), so it would be a much more simplified version because all the logic to group similar queries to a common denominator (which is the main reason why that dataloader is still out of tree) can be cut. I'd say it would be feasible to implement this, but I will have to look into the issue more in depth.

**Reply by darkbasic** [reply]
By the way that was my plan to gradually merge the find dataloader in the first place: to start by stripping out the query optimizer and just dataload the queries which match exactly.
It would still have its use cases and `em.populate` is one of them.

**Comment by wenerme** [comment]
@B4nan  Hi, I tried override PK like

```ts
@Entity({
 expression:`
   select distinct on (house_id) *
   from rent_contract
   order by house_id, end_date desc
 `
})
export class LastRentContractEntity extends RentContractEntity {
  @Property({ type: types.string })
  id!: string & Opt;

  @OneToOne(() => HouseEntity, { inversedBy: (v) => v.lastRentContract })
  house!: Rel<HouseEntity>;
}
```


but

```
TypeError: Cannot read properties of undefined (reading 'match')
    at EntityComparator.wrap (/opt/node_modules/.pnpm/@mikro-orm+core@6.2.0/node_modules/@mikro-orm/core/utils/EntityComparator.js:535:17)
    at propName (/opt/node_modules/.pnpm/@mikro-orm+core@6.2.0/node_modules/@mikro-orm/core/utils/EntityComparator.js:220:69)
    at /opt/node_modules/.pnpm/@mikro-orm+core@6.2.0/node_modules/@mikro-orm/core/utils/EntityComparator.js:297:44
    at Array.forEach (<anonymous>)
    at EntityComparator.getResultMapper (/opt/node_modules/.pnpm/@mikro-orm+core@6.2.0/node_modules/@mikro-orm/core/utils/EntityComparator.js:253:20)
    at EntityComparator.mapResult (/opt/node_modules/.pnpm/@mikro-orm+core@6.2.0/node_modules/@mikro-orm/core/utils/EntityComparator.js:45:29)
    at PostgreSqlDriver.mapResult (/opt/node_modules/.pnpm/@mikro-orm+core@6.2.0/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:67:32)
    at PostgreSqlDriver.mapResult (/opt/node_modules/.pnpm/@mikro-orm+knex@6.2.0_@mikro-orm+core@6.2.0_better-sqlite3@9.4.5_mysql2@3.9.4_pg@8.11.5_tedious@18.1.0/node_modules/@mikro-orm/knex/AbstractSqlDriver.js:160:27)
    at /opt/node_modules/.pnpm/@mikro-orm+knex@6.2.0_@mikro-orm+core@6.2.0_better-sqlite3@9.4.5_mysql2@3.9.4_pg@8.11.5_tedious@18.1.0/node_modules/@mikro-orm/knex/query/QueryBuilder.js:625:47
```

The code is

```js
    wrap(key) {
        if (key.match(/^\[.*]$/)) {
            return key;
        }
        return key.match(/^\w+$/) ? `.${key}` : `['${key}']`;
    }
```

The House side is defined as

```ts
  @OneToOne(() => LastRentContractEntity, { mappedBy: (v) => v.house })
  lastRentContract?: Rel<LastRentContractEntity>;
```

**Reply by B4nan** [reply]
This is not supported. You can't have a relation pointing to a virtual entity, just like you can't have PK on it. This is technically the same limitation, we just don't validate it (which I will most likely do). I told you to use a `formula` to build a virtual relation instead, that means the relation is represented by an SQL query, as such it cannot be bidirectional at all - but you can define another (virtual but owning) relation on the other side too, again with a `formula`.

**Reply by wenerme** [reply]
@B4nan  Thanks.

I think `formula ` do not support any form of dataloader if it's a FK, so I tried this first.

This formula way works, also dataloader works too, and keep `RentContractEntity` unchanged, but I think the **subquery** is not good.

```ts
  @Formula(
    (alias) => `(
select id 
from rent_contract 
where house_id = ${alias}.id 
order by end_date desc 
limit 1)`,
    {
      type: types.string,
    },
  )
  lastRentContractId?: string;

  @ManyToOne(() => RentContractEntity)
  lastRentContract?: Rel<RentContractEntity>;
```

I use a generic loader for this, but I don't know is this the right way to do so.

```ts
export function loadEntity(e: any) {
  if (!e) {
    return undefined;
  }
  if (e instanceof Reference) {
    if (e.isInitialized()) {
      return e.unwrap();
    }
    return e.load({ dataloader: true });
  }
  if (e instanceof BaseEntity) {
    if (e.isInitialized()) {
      return e;
    }
    return e.toReference().load({
      dataloader: true,
    });
  }
  return e;
}

```

One more little thing, the query log will replace formula `\n` to `''`, it should replace to `' '`

**Reply by B4nan** [reply]
You need the reference wrapper, you still don't have it there. It's required for dataloaders to work.

Also, this is wrong, I told you to use virtual relation, which means using `formula` option in the relation decorator. 

```ts
@ManyToOne(() => RentContractEntity, { formula: alias => '...', ref: true })
lastRentContract?: Ref<RentContractEntity>;
```

**Reply by B4nan** [reply]
Also, it makes no sense to do the explicit checks for `isInitailized` if you use `Reference.load` - that's exactly what the method is doing, loading only if its not loaded already.

**Reply by wenerme** [reply]
I know I missing the `ref: true`, but the strange part is this `e.toReference().load({ dataloader: true, })` seems works too, I want to avoid `ref: true` for better dev expr, but I really don't know should I do this ? Is this a undefined behavior or supported pattern ?

## Pattern 1: keep ID & Entity

Keep both Property, I hope I can

```ts
 @Formula(
    (alias) => `(
select id 
from rent_contract 
where house_id = ${alias}.id 
order by end_date desc 
limit 1)`,
    {
      type: types.string,
    },
  )
  lastRentContractId?: string;

  @ManyToOne(() => RentContractEntity)
  lastRentContract?: Rel<RentContractEntity>;
```

Otherwise I have to

```ts
  @ManyToOne({ entity: () => OpportunityEntity, nullable: true })
  opportunity?: OpportunityEntity;

  @Property({ type: types.string, nullable: true, persist: false })
  get opportunityId() {
    return this.opportunity?.id;
  }
```

## Pattern 2: Entity.toReference().load()

If I want dataloader I just explicit call load, this can dev experience.

```ts
e.toReference().load({
      dataloader: true, // omit this if  config on MikroORM
})
```

**Reply by wenerme** [reply]
Maybe we should have somthing like https://blog.axlight.com/posts/jotai-tips/ for advance pattern.
